google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,469 @@
+              } else {
+                should_finish = true;
+                stream_state_ = State::kRetrying;
+              }
+            }
+          }
+
+          if (should_finish) {
+            future<Status> finish_future;
+            {
+              std::lock_guard<std::mutex> g{mu_};
+              finish_future = stream_->Finish();
+            }
+            finish_future.then([this](future<Status> finish_status) {
+              std::lock_guard<std::mutex> g{mu_};
+              OnFailure(finish_status.get());
+            });
+          }
+
+          return response_good
+                     ? std::move(optional_response)
+                     : read_reinit_done_->get_future().then([](future<void>) {
+                         return make_ready_future(absl::optional());
+                       });
+        });
+  }
+
+  future<bool> Write(Request const& r, grpc::WriteOptions o) override {
+    future<bool> write_future;
+    bool retrying = false;
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      retrying = stream_state_ == State::kRetrying;
+      if (retrying) {
+        write_reinit_done_.emplace(absl::make_optional(promise<void>()));
+        write_future = write_reinit_done_->get_future();
+      } else {
+        write_future = stream_->Write(r, o);
+        outstanding_write_ = true;
+      }
+    }
+
+    if (retrying) {
+      return write_future;
+    }
+
+    return write_future.then([this](future<bool> write_fu) {
+      bool response_good = write_fu.get();
+      bool should_finish = false;
+
+      {
+        std::lock_guard<std::mutex> g{mu_};
+        outstanding_write_ = false;
+        if (stream_state_ == State::kMustRetry) {
+          should_finish = true;
+          stream_state_ = State::kRetrying;
+        }
+        if (!response_good) {
+          write_reinit_done_.emplace(absl::make_optional(promise<void>()));
+          if (outstanding_read_) {
+            stream_state_ = State::kMustRetry;
+          } else if (!should_finish) {
+            should_finish = true;
+            stream_state_ = State::kRetrying;
+          }
+        }
+      }
+
+      if (should_finish) {
+        future<Status> finish_future;
+        {
+          std::lock_guard<std::mutex> g{mu_};
+          finish_future = stream_->Finish();
+        }
+        finish_future.then([this](future<Status> finish_status) {
+          std::lock_guard<std::mutex> g{mu_};
+          OnFailure(finish_status.get());
+        });
+      }
+
+      return response_good
+                 ? make_ready_future(true)
+                 : write_reinit_done_->get_future().then(
+                       [](future<void>) { return make_ready_future(false); });
+    });
+  }
+
+  future<Status> Finish() override {
+    bool shutdown = false;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      shutdown = stream_state_ == State::kShutdown;
+      stream_state_ = State::kShutdown;
+    }
+
+    if (shutdown) {
```

Comment:
you can return with the mutex held- it'll release it since you use lock_guard.

```
{
  std::lock_guard<std::mutex> g{mu_};
  if (stream_state_ == State::kShutdown) return make_ready_future(kPermanentErrorStatus);
  stream_state_ = State::kShutdown;
}
```

Reply:
oh so the potential deadlock that you mentioned with `set_value` and Carlos mentioned with the `then` callback won't be seen with returning a simple future?

Reply:
nope. At that point, the lock is released because of the destructor of the lock guard

Reply:
ah yea RAII and scope